js/src/wrappers/ai-sdk/ai-sdk.ts

```diff
@@ -811,13 +827,50 @@ const wrapToolExecute = (tool: any, name: string) => {
     return new Proxy(tool, {
       get(target, prop) {
         if (prop === "execute") {
-          return (...args: any[]) =>
-            traced(
+          // Return a function that handles both regular async functions and async generators
+          const wrappedExecute = (...args: any[]) => {
+            const result = originalExecute.apply(target, args);
+
+            // Check if the result is an async generator (from async function* () {})
+            // AI SDK v5 supports async generator tools that yield intermediate status updates
+            if (isAsyncGenerator(result)) {
+              // Return a wrapper async generator that:
+              // 1. Iterates through the original generator
+              // 2. Yields all intermediate values (so consumers see status updates)
+              // 3. Tracks and logs the final yielded value as the tool output
+              return (async function* () {
+                const span = startSpan({
+                  name,
+                  spanAttributes: {
+                    type: SpanTypeAttribute.TOOL,
+                  },
+                });
+                span.log({ input: args.length === 1 ? args[0] : args });
+
+                try {
+                  let lastValue: any;
+                  for await (const value of result) {
+                    lastValue = value;
+                    yield value;
+                  }
+                  // Log the final yielded value as the output
+                  span.log({ output: lastValue });
+                  span.end();
```

Comment:
span.end() should be in `finally`, and then you can remove it from both the `try` and `catch` cases.

Reply:
Makes sense. I'll update.